File: July2022/C++/OutOfBoundaryPaths.cpp
// Recursion - TLE ❌
// TC: O(n^4)
// SC: O(m * n * moves)
class Solution {
private:
    int MOD = (int) (1e9 + 7);

    int helper(int i, int j, int moves, int m, int n) {
        // base case
        if (i < 0 || i == m || j < 0 || j == n) {
            return 1;
        }
        if (moves == 0) {
            return 0;
        }
        
        int top = helper(i - 1, j, moves - 1, m, n) % MOD;
        int down = helper(i + 1, j, moves - 1, m, n) % MOD;
        int left = helper(i, j - 1, moves - 1, m, n) % MOD;
        int right = helper(i, j + 1, moves - 1, m, n) % MOD;
        
        return (top + down) % MOD + (left + right) % MOD;
    }
    
public:    
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        return helper(startRow, startColumn, maxMove, m, n) % MOD;
    }
};



// Memoization - Accepted ✅
// TC: O(m * n * maxMove)
// SC: O(m * n * maxMove)   -> 3d dp array
class Solution {
private:
    int MOD = (int) (1e9 + 7);

    int helper(int i, int j, int moves, int m, int n, vector<vector<vector<int>>>& dp) {
        // base case
        if (i < 0 || i == m || j < 0 || j == n) {
            return 1;
        }
        if (moves == 0) {
            return 0;
        }

        if (dp[i][j][moves] != -1) {
            return dp[i][j][moves];
        }

        int top = helper(i - 1, j, moves - 1, m, n, dp) % MOD;
        int down = helper(i + 1, j, moves - 1, m, n, dp) % MOD;
        int left = helper(i, j - 1, moves - 1, m, n, dp) % MOD;
        int right = helper(i, j + 1, moves - 1, m, n, dp) % MOD;

        return dp[i][j][moves] = (top + down) % MOD + (left + right) % MOD;
    }
    
public:
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        vector<vector<vector<int>>> dp(m, vector<vector<int>>(n, vector<int>(maxMove + 1, -1)));        
        return helper(startRow, startColumn, maxMove, m, n, dp) % MOD;        
    }
};

File: July2022/Java/OutOfBoundaryPaths.java

// @saorav21994
// TC : O(m*n*maxMove)
// SC : O(m*n*maxMove)

class Solution {
    
    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        Integer [][][] memo = new Integer[m][n][maxMove+1];     // Default -> null
        return find(m, n, maxMove, startRow, startColumn, memo);
    }
    
    public int find(int m, int n, int move, int i, int j, Integer [][][] memo) {
        int mod = 1000000007;
        if (i < 0 || i >= m || j < 0 || j >= n) {
            return 1;       // we went outside the grid
        }
        if (move <= 0) {    // no more moves left
            return 0;
        }
        if (memo[i][j][move] != null) {
            return memo[i][j][move];    // memo already has this processed 
        }
        
        int count = 0;
        
        count = (count%mod) + (find(m, n, move-1, i-1, j, memo)%mod);
        count = (count%mod) + (find(m, n, move-1, i+1, j, memo)%mod);
        count = (count%mod) + (find(m, n, move-1, i, j-1, memo)%mod);
        count = (count%mod) + (find(m, n, move-1, i, j+1, memo)%mod);
        
        memo[i][j][move] = count%mod;
        
        return memo[i][j][move];
    }
}
